src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/EdifactParser.java

```diff
@@ -0,0 +1,157 @@
+
+    @Autowired
+    public EdifactParser(InterchangeFactory interchangeFactory) {
+        this.interchangeFactory = interchangeFactory;
+    }
+
+    public Interchange parse(String edifact) {
+        var allEdifactSegments = Arrays.asList(Split.bySegmentTerminator(edifact.replaceAll("\\n", "").strip()));
+
+        return parseInterchange(allEdifactSegments);
+    }
+
+    private Interchange parseInterchange(List<String> allEdifactSegments) {
+        Interchange interchange = interchangeFactory.createInterchange(
+                extractInterchangeEdifactSegments(allEdifactSegments));
+
+        var messages = parseAllMessages(allEdifactSegments);
+        messages.forEach(message -> message.setInterchange(interchange));
+        interchange.setMessages(messages);
+
+        return interchange;
+    }
+
+    private List<Message> parseAllMessages(List<String> allEdifactSegments) {
+        var allMessageHeaderSegmentIndexes = findAllIndexesOfSegment(allEdifactSegments, MessageHeader.KEY);
+        var allMessageTrailerSegmentIndexes = findAllIndexesOfSegment(allEdifactSegments, MessageTrailer.KEY);
+
+        var messageHeaderTrailerIndexPairs = zipIndexes(allMessageHeaderSegmentIndexes, allMessageTrailerSegmentIndexes);
+
+        return messageHeaderTrailerIndexPairs.stream()
+                .map(messageStartEndIndexPair ->
+                        allEdifactSegments.subList(messageStartEndIndexPair.getLeft(), messageStartEndIndexPair.getRight() + 1))
+                .map(this::parseMessage)
+                .collect(Collectors.toList());
+    }
+
+    private Message parseMessage(List<String> singleMessageEdifactSegments) {
+        var messageTrailerIndex = singleMessageEdifactSegments.size() - 1;
+        var firstTransactionStartIndex = findAllIndexesOfSegment(singleMessageEdifactSegments, TRANSACTION_START_SEGMENT).stream()
+                .findFirst()
+                // there might be no transaction inside - RECEP - so all message lines belong to the message
+                .orElse(messageTrailerIndex);
+
+        var onlyMessageLines = new ArrayList<>(singleMessageEdifactSegments.subList(0, firstTransactionStartIndex)); // first lines until transaction
+        onlyMessageLines.add(singleMessageEdifactSegments.get(messageTrailerIndex));
+
+        var message = new Message(onlyMessageLines);
+        var transactions = parseAllTransactions(singleMessageEdifactSegments);
+        transactions.forEach(transaction -> transaction.setMessage(message));
+        message.setTransactions(transactions);
+
+        return message;
+    }
+
+    private List<Transaction> parseAllTransactions(List<String> singleMessageEdifactSegments) {
+        var transactionStartIndexes = findAllIndexesOfSegment(singleMessageEdifactSegments, TRANSACTION_START_SEGMENT);
+        var transactionEndIndexes = new ArrayList<>(transactionStartIndexes);
+
+        // there might be no transactions inside - RECEP
```

Comment:
I'm slightly in favour of leaving the comments as they are because I wouldn't fully understand the functionality without them if I were new to the code base

Reply:
Generally we should avoid comments, and if we need to add comments, then the code is difficult to understand and needs refactored. But I understand the parsing/mapping logic is quite complex so comments may help with that (as long as the comments are understood by all).